apps/module-scan/src/util/qrcode.ts

```diff
@@ -82,70 +83,60 @@ export function* getSearchAreas(
 
 export const detectQRCode = async (
   imageData: ImageData
-): Promise<{ data: Buffer; position: 'top' | 'bottom' } | undefined> => {
-  const timer = time('detectQRCode')
-  let result: { data: Buffer; position: 'top' | 'bottom' } | undefined
-
+): Promise<
+  { data: Buffer; position: 'top' | 'bottom'; detector: string } | undefined
+> => {
   debug('detectQRCode: checking %dˣ%d image', imageData.width, imageData.height)
 
-  for (const { position, bounds } of getSearchAreas(imageData)) {
-    debug('cropping %s to check for QR code: %o', position, bounds)
-    const cropped = crop(imageData, bounds)
-    debug('scanning with qrdetect')
-    const results = qrdetect(cropped.data, cropped.width, cropped.height)
-
-    if (results.length > 0) {
-      debug(
-        'qrdetect found QR code in %s! data length=%d',
-        position,
-        results[0].data.length
-      )
-      const data = maybeDecodeBase64(results[0].data)
-
-      result = { data, position }
-    } else {
-      debug('qrdetect found no QR codes in %s', position)
-    }
-
-    if (result) {
-      break
-    }
-  }
-
-  if (!result) {
-    for (const { position, bounds } of getSearchAreas(imageData)) {
-      debug('cropping %s to check for QR code: %o', position, bounds)
-      const cropped = crop(imageData, bounds)
-
-      debug('generating PNG for quirc')
-      const img = await toPNG(cropped)
-
-      debug('scanning with quirc')
-      const results = (await quircDecode(img))[0]
-
-      if (results && 'data' in results) {
-        debug(
-          'quirc found QR code in %s! data length=%d',
-          position,
-          results.data.length
-        )
-        const data = maybeDecodeBase64(results.data)
-
-        result = { data, position }
-      } else {
-        debug('quirc found no QR codes in %s', position)
-      }
+  const detectors = [
+    {
+      name: 'qrdetect',
+      detect: async ({ data, width, height }: ImageData): Promise<Buffer[]> =>
+        qrdetect(data, width, height).map((symbol) => symbol.data),
+    },
+    {
+      name: 'quirc',
+      detect: async (imageData: ImageData): Promise<Buffer[]> => {
+        const results = await quircDecode(await toPNG(imageData))
+        return results
+          .filter((result): result is QRCode => !('err' in result))
+          .map((result) => result.data)
+      },
+    },
+    {
+      name: 'jsQR',
+      detect: async ({ data, width, height }: ImageData): Promise<Buffer[]> => {
+        const result = jsQR(data, width, height)
+        return result ? [Buffer.from(result.binaryData)] : []
+      },
+    },
+  ]
 
-      if (result) {
-        break
+  const timer = time('detectQRCode')
```

Comment:
not really in scope of this PR but just curious, do we have to keep up with the speed that the scanner physically scans things? Does it wait for our code to run between each scan?

Reply:
The scanners are capable of scanning faster than we currently process images, and we do make the scanners wait until we're done so that whoever is running the scanner can just take the top sheet out when there's a problem with it.